### PR TITLE
mon: break down raw usage for 'ceph df' by device class

### PR DESCRIPTION
**Pull request by liewegas**

### CODE REVIEW

File: src/mon/PGMap.cc [hunk 1]
```diff
@@ -870,8 +870,8 @@ void PGMapDigest::dump_cluster_stats(stringstream *ss,
     TextTable tbl;
     tbl.define_column("SIZE", TextTable::LEFT, TextTable::RIGHT);
     tbl.define_column("AVAIL", TextTable::LEFT, TextTable::RIGHT);
-    tbl.define_column("RAW USED", TextTable::LEFT, TextTable::RIGHT);
-    tbl.define_column("%RAW USED", TextTable::LEFT, TextTable::RIGHT);
+    tbl.define_column("USED", TextTable::LEFT, TextTable::RIGHT);
```
Comment (tchaikov) [inline src/mon/PGMap.cc]: might want to updated https://github.com/ceph/ceph/blob/master/doc/rados/operations/monitoring.rst accordingly. search for `RAW USED` in it.

File: src/mon/PGMap.cc [hunk 2]
```diff
@@ -861,27 +876,46 @@ void PGMapDigest::dump_cluster_stats(stringstream *ss,
...
+	  << stringify(byte_u_t(i.second.kb_avail*1024))
+	  << stringify(byte_u_t(i.second.kb_used*1024));
+      float used = 0.0;
+      if (i.second.kb > 0) {
+	used = ((float)i.second.kb_used / i.second.kb);
+      }
+      tbl << percentify(used*100);
+      tbl << TextTable::endrow;
     }
+    tbl << "TOTAL";
```
Comment (tchaikov) [inline src/mon/PGMap.cc]: might want to updated https://github.com/ceph/ceph/blob/master/doc/rados/operations/monitoring.rst accordingly. search for `OBJECTS` in it.

File: src/mon/PGMap.cc [hunk 3]
```diff
@@ -884,7 +884,7 @@ void PGMapDigest::dump_cluster_stats(stringstream *ss,
     }
     tbl << TextTable::endrow;
 
-    *ss << "GLOBAL:\n";
+    *ss << "RAW:\n";
```
Comment (tchaikov) [inline src/mon/PGMap.cc]:
```
2018-11-25T11:18:33.771 INFO:tasks.workunit.client.0.smithi008.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:723: test_mon_misc:  ceph df
2018-11-25T11:18:34.081 INFO:tasks.workunit.client.0.smithi008.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:724: test_mon_misc:  grep GLOBAL /tmp/cephto
ol.7vs/test_invalid.wT4
2018-11-25T11:18:34.082 INFO:tasks.workunit.client.0.smithi008.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:1: test_mon_misc:  rm -fr /tmp/cephtool.7vs
2018-11-25T11:18:34.083 DEBUG:teuthology.orchestra.run:got remote process result: 1
```

might want to update `qa/workunits/cephtool/test.sh` accordingly.

File: src/mon/PGMap.cc [hunk 4]
```diff
@@ -859,30 +876,49 @@ void PGMapDigest::dump_fs_stats(stringstream *ss, Formatter *f, bool verbose) co
     f->dump_int("total_used_bytes", osd_sum.kb_used * 1024ull);
     f->dump_int("total_avail_bytes", osd_sum.kb_avail * 1024ull);
     f->dump_float("total_percent_used", used * 100);
-    if (verbose) {
-      f->dump_int("total_objects", pg_sum.stats.sum.num_objects);
```
Comment (tchaikov) [inline src/mon/PGMap.cc]:
```
2018-11-25T11:08:13.861 INFO:teuthology.orchestra.run.smithi046.stderr:2018-11-25 11:08:13.859 7f3cc2ffd700 20 mgrc handle_command_reply command_reply(tid 0: -1 Test failed: Remote
method threw exception: Traceback (most recent call last):
2018-11-25T11:08:13.861 INFO:teuthology.orchestra.run.smithi046.stderr:  File "/usr/lib64/ceph/mgr/prometheus/module.py", line 608, in self_test
2018-11-25T11:08:13.861 INFO:teuthology.orchestra.run.smithi046.stderr:    self.collect()
2018-11-25T11:08:13.861 INFO:teuthology.orchestra.run.smithi046.stderr:  File "/usr/lib64/ceph/mgr/prometheus/module.py", line 514, in collect
2018-11-25T11:08:13.861 INFO:teuthology.orchestra.run.smithi046.stderr:    self.get_df()
2018-11-25T11:08:13.861 INFO:teuthology.orchestra.run.smithi046.stderr:  File "/usr/lib64/ceph/mgr/prometheus/module.py", line 304, in get_df
2018-11-25T11:08:13.862 INFO:teuthology.orchestra.run.smithi046.stderr:    self.metrics['cluster_{}'.format(stat)].set(df['stats'][stat])
2018-11-25T11:08:13.862 INFO:teuthology.orchestra.run.smithi046.stderr:KeyError: ('total_objects',)
```

see http://pulpito.ceph.com/kchai-2018-11-25_10:39:02-rados-wip-kefu2-testing-2018-11-22-1635-distro-basic-smithi/3282143/

**Comment by tchaikov** [comment]
this change conflicts with #19454. dropping from test batch.

File: src/mon/PGMap.cc [hunk 5]
```diff
@@ -1408,6 +1435,14 @@ void PGMap::calc_purged_snaps()
...
 }
 
+void PGMap::calc_osd_sum_by_class(const OSDMap& osdmap)
+{
+  osd_sum_by_class.clear();
+  for (auto& i : osd_stat) {
+    osd_sum_by_class[osdmap.crush->get_item_class(i.first)].add(i.second);
+  }
+}
+
```
Comment (tchaikov) [inline src/mon/PGMap.cc]:
```
ceph version 14.0.1-1539-gd5ab623 (d5ab6237de9307f957322527969f31b01ec0308e) nautilus (dev)
 1: (()+0x12890) [0x7f0b7bd89890]
 2: (gsignal()+0xc7) [0x7f0b7ae7ce97]
 3: (abort()+0x141) [0x7f0b7ae7e801]
 4: (()+0x8c8fb) [0x7f0b7b8718fb]
 5: (()+0x92d3a) [0x7f0b7b877d3a]
 6: (()+0x92d95) [0x7f0b7b877d95]
 7: (()+0x92fe8) [0x7f0b7b877fe8]
 8: (()+0x8e803) [0x7f0b7b873803]
 9: (()+0xc9b7e) [0x562cd8cb9b7e]
 10: (PGMap::calc_osd_sum_by_class(OSDMap const&)+0x118) [0x562cd8cc9a28]
 11: (PGMap::encode_digest(OSDMap const&, ceph::buffer::list&, unsigned long)+0x2a) [0x562cd8cd4cfa]
 12: (DaemonServer::send_report()+0x52c) [0x562cd8d6ad5c]
 13: (DaemonServer::maybe_ready(int)+0x44e) [0x562cd8d6c06e]
 14: (DaemonServer::ms_dispatch(Message*)+0x1b7) [0x562cd8d6c417]
 15: (Dispatcher::ms_dispatch2(boost::intrusive_ptr<Message> const&)+0x26) [0x562cd8d7e236]
 16: (DispatchQueue::entry()+0x1489) [0x7f0b7cc01969]
 17: (DispatchQueue::DispatchThread::entry()+0xd) [0x7f0b7ccb378d]
 18: (()+0x76db) [0x7f0b7bd7e6db]
```

/a/kchai-2018-12-09_00:37:50-rados-wip-kefu2-testing-2018-12-09-0002-distro-basic-smithi/3318978/remote/smithi151/log/mgr.x.log

**Comment by sebastian-philipp** [comment]
This broke the dashboard qa tests.

**Pull request merged**